quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,830 @@
+    match (func, simplify(expr, limit.saturating_sub(1))) {
+        (ExpressionFunction::Cis, Expression::Number(x)) => {
+            // num_complex::Complex64::cis only accpets f64 :-(
+            Expression::Number(x.cos() + I * x.sin())
+        }
+        (ExpressionFunction::Cis, Expression::PiConstant) => Expression::Number(-ONE),
+        (ExpressionFunction::Cosine, Expression::Number(x)) => Expression::Number(x.cos()),
+        (ExpressionFunction::Cosine, Expression::PiConstant) => Expression::Number(-ONE),
+        (ExpressionFunction::Exponent, Expression::Number(x)) => Expression::Number(x.exp()),
+        (ExpressionFunction::Exponent, Expression::PiConstant) => Expression::Number(PI.exp()),
+        (ExpressionFunction::Sine, Expression::Number(x)) => Expression::Number(x.sin()),
+        (ExpressionFunction::Sine, Expression::PiConstant) => Expression::Number(PI.sin()),
+        (ExpressionFunction::SquareRoot, Expression::Number(x)) => Expression::Number(x.sqrt()),
+        (ExpressionFunction::SquareRoot, Expression::PiConstant) => Expression::Number(PI.sqrt()),
+        (function, expression) => Expression::FunctionCall(FunctionCallExpression {
+            function,
+            expression: expression.into(),
+        }),
+    }
+}
+
+#[inline]
+fn is_zero(x: num_complex::Complex64) -> bool {
+    x.norm() < 1e-10
+}
+
+#[inline]
+fn is_one(x: num_complex::Complex64) -> bool {
+    is_zero(x - 1.0)
+}
+
+/// Helper: in simplification, we'll bias towards smaller expressions
+fn size(expr: &Expression) -> usize {
+    match expr {
+        Expression::Address(_)
+        | Expression::Number(_)
+        | Expression::PiConstant
+        | Expression::Variable(_) => 1,
+        Expression::FunctionCall(FunctionCallExpression {
+            function: _,
+            expression,
+        }) => 1 + size(expression),
+        Expression::Infix(InfixExpression {
+            left,
+            operator: _,
+            right,
+        }) => 1 + size(left) + size(right),
+        Expression::Prefix(PrefixExpression {
+            operator: _,
+            expression,
+        }) => 1 + size(expression),
+    }
+}
+
+// It's verbose to go alone! Take this.
+macro_rules! infix {
+    ($left:expr, $op:expr, $right:expr) => {
+        Expression::Infix(InfixExpression {
+            left: $left.into(),
+            operator: $op,
+            right: $right.into(),
+        })
+    };
+}
+macro_rules! add {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Plus, $right)
+    };
+}
+macro_rules! sub {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Minus, $right)
+    };
+}
+macro_rules! mul {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Star, $right)
+    };
+}
+macro_rules! div {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Slash, $right)
+    };
+}
+
+// Are these both of the form something * x for the _same_ x?
```

Comment:
Minor thing: this initially reads to me like a TODO, not an explanation of what the function does.

```suggestion
// Checks if both ar eof the form something * x for the _same_ x.
```

Reply:
Would be a good idea to leave proper doc comments as well, for future people coming to this.